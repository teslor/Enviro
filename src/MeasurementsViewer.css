.measurements-viewer-header {
  display: grid;
  grid-gap: 3px;
  grid-template-columns: repeat(5, 20%);
  background-color: #fff;
  color: #444;
  font-size: 1em;
  margin-bottom: 5px;
}

.measurements-viewer-header .box {
  background-color: #444;
  color: #fff;
  border-radius: 10px;
  padding: 12px;
  text-align: center;
  text-transform: uppercase;
}

.measurements-viewer-header .header {
  background-color: #00a200;
  grid-column: 2 / 5;
}

.measurements-viewer-header .button {
  background-color: #7ca7ff;
  cursor: pointer;
}

.measurements-viewer-content {
  display: grid;
  grid-gap: 4px;
  grid-template-columns: 35% 15% 25% 25%;
  background-color: #fff;
  color: #444;
  font-size: 1em;
}

.measurements-viewer-content .box {
  background-color: #999;
  color: #fff;
  border-radius: 5px;
  padding: 8px;
  font-size: 1em;
}

.measurements-viewer-content .header {
  background-color: #676767;
  padding: 12px;
  text-align: center;
  text-transform: uppercase;
}

.measurements-viewer-content .info {
  background-color: #999;
  grid-column: 1 / span 4;
  padding: 24px;
  text-align: center;
  text-transform: uppercase;
}

.measurements-viewer-content .a-center {
  text-align: center;
}

.measurements-viewer-content .a-right {
  text-align: right;
}