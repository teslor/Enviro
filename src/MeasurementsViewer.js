import React from 'react';
import cloneDeep from 'lodash/cloneDeep';
import throttle from 'lodash/throttle';
import './MeasurementsViewer.css';

const STATUS = {
  ONLINE: 'Online',
  OFFLINE: 'Offline',
  CONNECTING: 'Connecting',
};

/**
 * MeasurementsViewer connects to event source & shows received data/measurements in the table
 */
class MeasurementsViewer extends React.Component {
  constructor(props) {
    super(props);

    /**
		 * status - connection status
		 * items - measurements to show
		 */
    this.state = {
			status: STATUS.OFFLINE,
			items: [],
    };
  }

	/**
	 * Connects/disconnects to event source & sets event handlers
	 */
  toggleConnection() {
  	if (this.state.status === STATUS.CONNECTING) return;

  	if (!this.connection) {
  		if (!this.props.source) return;

			this.connection = new EventSource(this.props.source);

			// Set connection handlers

			const refreshRate = this.props.refreshRate ? this.props.refreshRate : 100;
			this.connection.onmessage = throttle(this.processMessage.bind(this), refreshRate);

			this.connection.onopen = () => {
				this.setState({ status: STATUS.ONLINE });
			};

			this.connection.onerror = (e) => {
				this.setState({ status: STATUS.OFFLINE });
			};

			this.setState({ status: STATUS.CONNECTING });
		} else {
  		this.connection.close();
  		this.connection = null;
  		this.setState({ status: STATUS.OFFLINE });
		}
	}

	/**
	 * Performs messages handling from event source & updates measurements to show
	 */
  processMessage(msg) {
		let data;

		try {
			data = JSON.parse(msg.data);
		} catch(err) {
			return;
		}

		let currentItems = cloneDeep(this.state.items);
		data.forEach((item) => {
			if (!item.name || !item.measurements || item.measurements.length < 1) {  // no name or measurements for this item
				return;
			}
			let lastMeasurement = item.measurements[item.measurements.length -  1];  //take the latest measurement

			let i, newItem;
			for (i = 0; i < currentItems.length; i += 1) {
				if (currentItems[i].data[0] !== item.name) {
					continue;
				}
				break;
			}
			if (i === currentItems.length) {
				newItem = {};
				currentItems.push(newItem);
			} else {
				newItem = currentItems[i];
			}

			newItem._id = item._id;
			newItem.data = [];
			newItem.data.push(item.name);
			newItem.data.push(item.unit);

			// TODO: use date formatting function if necessary
			// Time
			let time = new Date(lastMeasurement[0] * 1000).toString();
			time = time.match(/.+?(?= GMT)/)[0];
			newItem.data.push(time);

			// Value
			let value = lastMeasurement[1];
			switch(typeof value) {
				case 'number':
					value = value.toFixed(2);
					break;
				case 'object':  // location
					value = `${value[0]} ${value[1]}`;
					break;
				default:
			}
			newItem.data.push(value);
		});

		this.setState({ items: currentItems });
	}

	handleSwitchClick = () => {
		this.toggleConnection();
	};

  render() {
  	let content = [];
  	this.state.items.forEach((item) => {
  		item.data.forEach((dataItem, i) => {
  			let classes = 'box content';
  			let key = item._id;

				switch(i) {
					case 0:
						key = dataItem;
						break;
					case 1:
						classes += ' a-center';
						key += '1';
						break;
					case 2:
						classes += ' a-center';
						key += '2';
						break;
					case 3:
						classes += ' a-right';
						key += '3';
						break;
					default:
				}

				content.push(
					<div className={classes} key={key}>{dataItem}</div>
				)
			})
		});

  	if (!content.length) {
  	  content.push(
				<div className='box info' key='info'>No measurements to show</div>
			)
		}

		//TODO: possible improvement: color highlighting for values which require attention
    return (
    	<div className="measurements-viewer">
				<div className="measurements-viewer-header">
					<div className="box button" onClick={this.handleSwitchClick}>
						Turn {
						  this.state.status === STATUS.ONLINE ? 'Off' : (this.state.status === STATUS.OFFLINE ? 'On' : 'On...')
					  }
					</div>
					<div className="box header">Measurements</div>
					<div className="box button">{this.state.status}</div>
				</div>
				<div className="measurements-viewer-content">
					<div className="box header">Name</div>
					<div className="box header">Unit</div>
					<div className="box header">Time</div>
					<div className="box header">Value</div>
					{content}
				</div>
			</div>
    );
	}
}

export default MeasurementsViewer;