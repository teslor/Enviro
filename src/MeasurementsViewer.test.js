import React from 'react';
import ReactDOM from 'react-dom';
import MeasurementsViewer from './MeasurementsViewer';
import renderer from 'react-test-renderer';

// const fixture = 'data: [{"name":"Batt. Voltage","unit":"V","measurements":[],"_id":"5971e022c0d1420001edb3bd"}]';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MeasurementsViewer />, div);
});

it('renders correctly', () => {
  const tree = renderer.create(
    <MeasurementsViewer />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

//TODO: add more tests