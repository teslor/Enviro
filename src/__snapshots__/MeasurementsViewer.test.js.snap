// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="measurements-viewer"
>
  <div
    className="measurements-viewer-header"
  >
    <div
      className="box button"
      onClick={[Function]}
    >
      Turn 
      On
    </div>
    <div
      className="box header"
    >
      Measurements
    </div>
    <div
      className="box button"
    >
      Offline
    </div>
  </div>
  <div
    className="measurements-viewer-content"
  >
    <div
      className="box header"
    >
      Name
    </div>
    <div
      className="box header"
    >
      Unit
    </div>
    <div
      className="box header"
    >
      Time
    </div>
    <div
      className="box header"
    >
      Value
    </div>
    <div
      className="box info"
    >
      No measurements to show
    </div>
  </div>
</div>
`;
